### PR TITLE
Add nested object return support for tracing autograd w/ grad

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105103

Add tuple return support for tracing autograd w/ grad

cc @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

### CODE REVIEW

File: torch/_dynamo/variables/higher_order_ops.py [hunk 1]
```diff
@@ -665,9 +674,22 @@ def call_function(
...
+                if isinstance(item, TensorVariable):
+                    return item.as_proxy().node.meta["example_value"].contiguous()
+                if item.has_unpack_var_sequence(tx):
+                    r_items = []
+                    for sub_item in item.unpack_var_sequence(tx):
+                        r_items.append(_unpack(sub_item))
+                    return item.python_type()(r_items)
+                raise RuntimeError("Unsupported autograd w/ grad return type.")
+
+            example_value = _unpack(arg)
```
Comment (voznesenskym) [inline torch/_dynamo/variables/higher_order_ops.py:691]: TODO: Util this if directionally folks don't hate it (cc @zou3519)

**Comment by voznesenskym** [comment]
Test failures look real, looking. Direction remains the same.

File: test/dynamo/test_autograd_function.py [hunk 2]
```diff
@@ -214,6 +214,24 @@ def forward(self, x):
...
+    # Test tuple return
+    @staticmethod
+    def forward(ctx, foo):
+        result = foo + foo
+        result = result + foo
+        return (result, result, result)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        return grad_output
```
Comment (zou3519) [inline test/dynamo/test_autograd_function.py:227]: backward should accept a number of grad_output equal to the number of outputs (3)

File: torch/_dynamo/variables/higher_order_ops.py [hunk 3]
```diff
@@ -772,8 +797,8 @@ def call_function(
             *(arg.as_proxy() for arg in args),
             *(arg for arg in body_lifted_freevars.keys()),
         )
-        r = body_r.as_proxy().node.meta["example_value"]
-        example_value = r
+
+        example_value = _unpack(tx, body_r)
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:801]: We actually already handle this non-tuple returns for `wrap` and checkpoint: https://github.com/pytorch/pytorch/blob/053654b9cf9ccb7f4f1746e7d324d694ce98f9f1/torch/_dynamo/variables/higher_order_ops.py#L826-L831 . It looks like we never brought the improvement over to autograd.Function. We should probably be consistent with how we solve the problem. 

I'm not completely sure what the difference is between `_unpack` and what we do with wrap, but we should settle on one and use it everywhere

File: torch/_dynamo/variables/higher_order_ops.py [hunk 4]
```diff
@@ -57,43 +57,59 @@ def are_tensors(var):
...
-            new_arg = a
+                tracer.create_graph_input(subarg.as_proxy().node.name)
+            new_arg = subarg
+        elif not isinstance(
+            subarg, NNModuleVariable
+        ) and subarg.has_unpack_var_sequence(tx):
+            new_arg = []
+            for sub_a in subarg.unpack_var_sequence(tx):
+                new_arg.append(add_subarg(sub_a))
+            new_arg = subarg
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:104]: Why do we need this?

If the problem is that HigherOrderOperator body functions don't accept pytree inputs (e.g. a TensorList), the current thinking was that:
- we would not actually record any inputs to the subgraph in this function
- Tensors will automatically get lifted to being an input to the subgraph using the free variable lifting mechanism

This is what the `manually_set_subgraph_inputs=False` flag does (but we'd also need to relax the constraints on the input types, which we haven't done yet)

Comment (zou3519) [reply torch/_dynamo/variables/higher_order_ops.py:104]: Actually I need to think about this some more...

File: test/dynamo/test_autograd_function.py (hunk 2 above)
Comment (voznesenskym) [reply test/dynamo/test_autograd_function.py:227]: got it

File: torch/_dynamo/variables/higher_order_ops.py (hunk 4 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/higher_order_ops.py:104]: Sorry, say more?

File: torch/_dynamo/variables/higher_order_ops.py (hunk 3 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/higher_order_ops.py:801]: They are subtly different - but I think we can pick one for our purposes. I just didn't know this existed! Let me try reusing this. 

> It looks like we never brought the improvement over to autograd.Function

Or functorch, it seems?

**Pull request closed without merging**